File: mtg-tools/frontend/src/local/operations.ts
/**
 * The undo log — port of `webapp/operations.py`, semantics preserved line for
 * line (including error messages: the UI shows them, and the parity gate
 * diffs them).
 *
 * The model is an **inverse patch**: every mutating action records, in the
 * same transaction, the row state needed to put things back. Undo replays
 * that state. It stores only the rows an operation touched, which keeps a
 * 400-row bulk edit small and exact.
 */

import { now, type Database } from './db'

/** Only these tables participate. Anything else must be reconstructible from
 * them, so an operation can never leave a half-reversible trail. */
export const UNDOABLE_TABLES = [
  'holdings',
  'sealed',
  'verdicts',
  'sales',
  'imports',
  'staged_rows',
] as const

export type RowKey = number | string | Array<number | string>
export type RowDict = Record<string, unknown>

export interface OperationRecord {
  id: number
  kind: string
  summary: string
  affected: number
  createdAt: string
  revertedAt: string | null
  reverted: boolean
}

/** Undo's "can't do that" — maps to HTTP 409 at the endpoint edge. */
export class UndoLookupError extends Error {}

function toRecord(row: RowDict): OperationRecord {
  return {
    id: row.id as number,
    kind: row.kind as string,
    summary: row.summary as string,
    affected: row.affected as number,
    createdAt: row.created_at as string,
    revertedAt: (row.reverted_at as string | null) ?? null,
    reverted: row.reverted_at !== null && row.reverted_at !== undefined,
  }
}

/** Primary key columns. `verdicts` is the one composite key. */
function pk(table: string): string[] {
  return table === 'verdicts' ? ['subject_kind', 'subject_id'] : ['id']
}

function assertUndoable(table: string): void {
  if (!(UNDOABLE_TABLES as readonly string[]).includes(table)) {
    throw new Error(`${table} is not undoable`)
  }
}

/**
 * Capture rows exactly as they are now, for later restoration. Call this
 * *before* mutating. Rows that don't exist yet simply aren't captured, which
 * is what makes an insert reverse into a delete.
 */
export function snapshotRows(db: Database, table: string, ids: RowKey[]): RowDict[] {
  assertUndoable(table)
  const keys = pk(table)
  if (!ids.length) return []

  if (keys.length === 1) {
    const placeholders = ids.map(() => '?').join(',')
    return db.selectObjects(
      `SELECT * FROM ${table} WHERE ${keys[0]} IN (${placeholders})`,
      ids as unknown[],
    )
  }

  const rows: RowDict[] = []
  for (const keyTuple of ids) {
    const where = keys.map((k) => `${k} = ?`).join(' AND ')
    const found = db.selectObject(
      `SELECT * FROM ${table} WHERE ${where}`,
      keyTuple as unknown[],
    )
    if (found !== undefined) rows.push(found)
  }
  return rows
}

/**
 * Write one undo entry. Must be called inside the mutation's transaction.
 *
 * `before`  — {table: [row dicts as they were]}; restored on undo.
 * `created` — {table: [primary keys that did not exist before]}; deleted on undo.
 *
 * A row that appears in both is handled correctly: `created` deletes run
 * first, then `before` restores, so an update that also inserted reverses
 * cleanly.
 */
export function record(
  db: Database,
  kind: string,
  summary: string,
  options: {
    before?: Record<string, RowDict[]>
    created?: Record<string, RowKey[]>
    affected?: number
  } = {},
): number {
  const inverse = {
    before: options.before ?? {},
    created: Object.fromEntries(
      Object.entries(options.created ?? {}).map(([table, keys]) => [
        table,
        keys.map((k) => (Array.isArray(k) ? [...k] : k)),
      ]),
    ),
  }
  for (const table of [...Object.keys(inverse.before), ...Object.keys(inverse.created)]) {
    assertUndoable(table)
  }

  db.exec({
    sql:
      'INSERT INTO operations (kind, summary, affected, inverse, created_at) ' +
      'VALUES (?, ?, ?, ?, ?)',
    bind: [kind, summary, options.affected ?? 0, JSON.stringify(inverse), now()],
  })
  return Number(db.selectValue('SELECT last_insert_rowid()'))
}

function resolveTarget(db: Database, operationId?: number): OperationRecord {
  const newest = db.selectObject(
    'SELECT * FROM operations WHERE reverted_at IS NULL ORDER BY id DESC LIMIT 1',
  )
  if (newest === undefined) throw new UndoLookupError('nothing to undo')
  if (operationId === undefined) return toRecord(newest)

  const row = db.selectObject('SELECT * FROM operations WHERE id = ?', [operationId])
  if (row === undefined) throw new UndoLookupError(`no operation ${operationId}`)
  if (row.reverted_at !== null && row.reverted_at !== undefined) {
    throw new UndoLookupError(`operation ${operationId} is already undone`)
  }
  if (row.id !== newest.id) {
    throw new UndoLookupError(
      `operation ${operationId} is not the most recent change ` +
        `(that is #${newest.id}: ${newest.summary}). ` +
        'Undo works newest-first so the result always matches the log.',
    )
  }
  return toRecord(row)
}

/**
 * Reverse one operation. Defaults to the most recent un-reverted one.
 *
 * Deliberately strict: only the newest un-reverted operation can be undone.
 * Undoing out of order would silently produce a state neither the user nor
 * the log describes.
 */
export function undoOperation(db: Database, operationId?: number): OperationRecord {
  const target = resolveTarget(db, operationId)
  const inverse = JSON.parse(
    db.selectValue('SELECT inverse FROM operations WHERE id = ?', [target.id]) as string,
  ) as { before?: Record<string, RowDict[]>; created?: Record<string, RowKey[]> }

  // Deletes first: an operation that both created and updated rows reverses
  // cleanly only in this order.
  for (const [table, keys] of Object.entries(inverse.created ?? {})) {
    const columns = pk(table)
    for (const key of keys) {
      const values = Array.isArray(key) ? key : [key]
      const where = columns.map((c) => `${c} = ?`).join(' AND ')
      db.exec({ sql: `DELETE FROM ${table} WHERE ${where}`, bind: values })
    }
  }

  for (const [table, rows] of Object.entries(inverse.before ?? {})) {
    for (const row of rows) {
      const columns = Object.keys(row)
      const placeholders = columns.map(() => '?').join(',')
      db.exec({
        sql:
          `INSERT OR REPLACE INTO ${table} (${columns.join(',')}) ` +
          `VALUES (${placeholders})`,
        bind: columns.map((c) => row[c] ?? null),
      })
    }
  }

  db.exec({
    sql: 'UPDATE operations SET reverted_at = ? WHERE id = ?',
    bind: [now(), target.id],
  })
  return target
}

export function recent(db: Database, limit = 25): OperationRecord[] {
  return db
    .selectObjects('SELECT * FROM operations ORDER BY id DESC LIMIT ?', [limit])
    .map(toRecord)
}

export function latestUndoable(db: Database): OperationRecord | null {
  const row = db.selectObject(
    'SELECT * FROM operations WHERE reverted_at IS NULL ORDER BY id DESC LIMIT 1',
  )
  return row === undefined ? null : toRecord(row)
}

File: mtg-tools/webapp/operations.py
"""The undo log.

Built before any mutation exists, deliberately. Retrofitting undo onto
mutations that already ship is how half of them end up unreversible — and bulk
edits over 543 rows are exactly where one misclick costs an afternoon.

The model is an **inverse patch**: every mutating action records, in the same
transaction, the row state needed to put things back. Undo replays that state.
It is not a generic "diff the whole database" scheme — it stores only the rows
an operation touched, which keeps a 400-row bulk edit small and exact.
"""

from __future__ import annotations

import json
import sqlite3
from typing import Any, Dict, Iterable, List, Optional, Sequence

from .db import now

__all__ = [
    "UNDOABLE_TABLES",
    "Operation",
    "snapshot_rows",
    "record",
    "undo",
    "recent",
    "latest_undoable",
]

#: Only these tables participate. Anything else must be reconstructible from
#: them, so an operation can never leave a half-reversible trail.
UNDOABLE_TABLES = ("holdings", "sealed", "verdicts", "sales", "imports", "staged_rows")


class Operation:
    """A recorded, reversible change."""

    __slots__ = ("id", "kind", "summary", "affected", "created_at", "reverted_at")

    def __init__(self, row: sqlite3.Row):
        self.id = row["id"]
        self.kind = row["kind"]
        self.summary = row["summary"]
        self.affected = row["affected"]
        self.created_at = row["created_at"]
        self.reverted_at = row["reverted_at"]

    @property
    def reverted(self) -> bool:
        return self.reverted_at is not None

    def __repr__(self) -> str:
        state = " (undone)" if self.reverted else ""
        return f"<Operation {self.id} {self.kind}: {self.summary}{state}>"


def _pk(table: str) -> Sequence[str]:
    """Primary key columns. `verdicts` is the one composite key."""
    if table == "verdicts":
        return ("subject_kind", "subject_id")
    return ("id",)


def snapshot_rows(
    conn: sqlite3.Connection, table: str, ids: Iterable
) -> List[Dict[str, Any]]:
    """Capture rows exactly as they are now, for later restoration.

    Call this *before* mutating. Rows that don't exist yet simply aren't
    captured, which is what makes an insert reverse into a delete.
    """
    if table not in UNDOABLE_TABLES:
        raise ValueError(f"{table} is not undoable")

    keys = _pk(table)
    ids = list(ids)
    if not ids:
        return []

    if len(keys) == 1:
        placeholders = ",".join("?" for _ in ids)
        rows = conn.execute(
            f"SELECT * FROM {table} WHERE {keys[0]} IN ({placeholders})", ids
        ).fetchall()
    else:
        rows = []
        for key_tuple in ids:
            where = " AND ".join(f"{k} = ?" for k in keys)
            found = conn.execute(
                f"SELECT * FROM {table} WHERE {where}", tuple(key_tuple)
            ).fetchone()
            if found is not None:
                rows.append(found)

    return [dict(r) for r in rows]


def record(
    conn: sqlite3.Connection,
    kind: str,
    summary: str,
    *,
    before: Optional[Dict[str, List[Dict[str, Any]]]] = None,
    created: Optional[Dict[str, List]] = None,
    affected: int = 0,
) -> int:
    """Write one undo entry. Must be called inside the mutation's transaction.

    `before`  — {table: [row dicts as they were]}; restored on undo.
    `created` — {table: [primary keys that did not exist before]}; deleted on undo.

    A row that appears in both is handled correctly: `created` deletes run first,
    then `before` restores, so an update that also inserted reverses cleanly.
    """
    inverse = {
        "before": before or {},
        "created": {k: [list(i) if isinstance(i, (list, tuple)) else i for i in v]
                    for k, v in (created or {}).items()},
    }
    for table in list(inverse["before"]) + list(inverse["created"]):
        if table not in UNDOABLE_TABLES:
            raise ValueError(f"{table} is not undoable")

    cursor = conn.execute(
        "INSERT INTO operations (kind, summary, affected, inverse, created_at) "
        "VALUES (?, ?, ?, ?, ?)",
        (kind, summary, affected, json.dumps(inverse), now()),
    )
    return cursor.lastrowid


def undo(conn: sqlite3.Connection, operation_id: Optional[int] = None) -> Operation:
    """Reverse one operation. Defaults to the most recent un-reverted one.

    Deliberately strict: only the newest un-reverted operation can be undone.
    Undoing out of order would silently produce a state neither the user nor the
    log describes — a bulk price edit undone *after* a later merge would restore
    prices onto rows the merge has since collapsed.
    """
    target = _resolve_target(conn, operation_id)
    inverse = json.loads(
        conn.execute(
            "SELECT inverse FROM operations WHERE id = ?", (target.id,)
        ).fetchone()["inverse"]
    )

    # Deletes first: an operation that both created and updated rows reverses
    # cleanly only in this order.
    for table, keys in inverse.get("created", {}).items():
        pk = _pk(table)
        for key in keys:
            values = key if isinstance(key, list) else [key]
            where = " AND ".join(f"{c} = ?" for c in pk)
            conn.execute(f"DELETE FROM {table} WHERE {where}", tuple(values))

    for table, rows in inverse.get("before", {}).items():
        for row in rows:
            columns = list(row)
            placeholders = ",".join("?" for _ in columns)
            conn.execute(
                f"INSERT OR REPLACE INTO {table} ({','.join(columns)}) "
                f"VALUES ({placeholders})",
                tuple(row[c] for c in columns),
            )

    conn.execute(
        "UPDATE operations SET reverted_at = ? WHERE id = ?", (now(), target.id)
    )
    return target


def _resolve_target(conn: sqlite3.Connection, operation_id: Optional[int]) -> Operation:
    newest = conn.execute(
        "SELECT * FROM operations WHERE reverted_at IS NULL ORDER BY id DESC LIMIT 1"
    ).fetchone()
    if newest is None:
        raise LookupError("nothing to undo")

    if operation_id is None:
        return Operation(newest)

    row = conn.execute(
        "SELECT * FROM operations WHERE id = ?", (operation_id,)
    ).fetchone()
    if row is None:
        raise LookupError(f"no operation {operation_id}")
    if row["reverted_at"] is not None:
        raise LookupError(f"operation {operation_id} is already undone")
    if row["id"] != newest["id"]:
        raise LookupError(
            f"operation {operation_id} is not the most recent change "
            f"(that is #{newest['id']}: {newest['summary']}). "
            f"Undo works newest-first so the result always matches the log."
        )
    return Operation(row)


def recent(conn: sqlite3.Connection, limit: int = 25) -> List[Operation]:
    rows = conn.execute(
        "SELECT * FROM operations ORDER BY id DESC LIMIT ?", (limit,)
    ).fetchall()
    return [Operation(r) for r in rows]


def latest_undoable(conn: sqlite3.Connection) -> Optional[Operation]:
    row = conn.execute(
        "SELECT * FROM operations WHERE reverted_at IS NULL ORDER BY id DESC LIMIT 1"
    ).fetchone()
    return Operation(row) if row else None
